### PR TITLE
feat: works with useLayoutEffect

### PR DESCRIPTION
**Pull request by leobastiani**
close https://github.com/pmndrs/zustand/issues/276

### CODE REVIEW

File: tests/basic.test.tsx [hunk 1]
```diff
@@ -147,6 +147,25 @@ it('only re-renders if selected state has changed', async () => {
...
+    React.useLayoutEffect(() => {
+      useStore.setState({ state: true })
+    }, [])
+    return <>{`${state}`}</>
+  }
+
+  const { container } = render(<Test />)
+
+  expect(container.innerHTML.trim()).toBe('true')
+  expect(useEffect).not.toHaveBeenCalled()
```
Comment (dai-shi) [inline tests/basic.test.tsx:166]: Hm, this is testing the implementation details. In the future we might be using `useEffect` for different purpose.
Can we test the behavior? How about adding useEffect in `Test()` and make sure it's called just once?

**Review by dai-shi — COMMENTED** [review]
Thanks for your work. A comment on test.

File: tests/basic.test.tsx (hunk 1 above)
Comment (leobastiani) [reply tests/basic.test.tsx:166]: After searching a lot, I actually found how to properly test this behavior. The problem with react-testing-library is that it flushes all the useEffects and useLayoutEffects. Now it is testing the behavior.

**Review by dai-shi — APPROVED** [review]
It looks good!

**Comment by leobastiani** [comment]
I just added a cleanup because we aren't using react-testing-library in this test

**Comment by dai-shi** [comment]
Nice.
Unless anyone else says anything, I'll merge this and release a patch version later.

**Comment by dai-shi** [comment]
Published: https://www.npmjs.com/package/zustand/v/3.3.1

**Pull request merged**